pkg/channel/distributed/controller/util/compare.go

```diff
@@ -0,0 +1,192 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"go.uber.org/zap"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"knative.dev/pkg/apis/duck"
+)
+
+// CheckDeploymentChanged Modifies An Existing Deployment With New Fields (If Necessary)
+// Returns True If Any Modifications Were Made
+func CheckDeploymentChanged(logger *zap.Logger, existingDeployment, newDeployment *appsv1.Deployment) (*appsv1.Deployment, bool) {
+
+	// Make a copy of the existing labels and annotations so we don't inadvertently
+	// modify the existing deployment fields directly
+	updatedLabels := make(map[string]string)
+	for oldKey, oldValue := range existingDeployment.ObjectMeta.Labels {
+		updatedLabels[oldKey] = oldValue
+	}
+	updatedAnnotations := make(map[string]string)
+	for oldKey, oldValue := range existingDeployment.Spec.Template.ObjectMeta.Annotations {
+		updatedAnnotations[oldKey] = oldValue
+	}
+
+	metadataChanged := false
+	// Add any labels in the "new" deployment to the copy of the labels from the old deployment.
+	for newKey, newValue := range newDeployment.ObjectMeta.Labels {
+		oldValue, ok := existingDeployment.ObjectMeta.Labels[newKey]
+		if !ok || oldValue != newValue {
+			metadataChanged = true
+			updatedLabels[newKey] = newValue
+		}
+	}
+
+	// Add any annotations in the "new" deployment to the copy of the labels from the old deployment.
+	// (In particular this will trigger on differences in "kafka.eventing.knative.dev/configmap-hash")
+	for newKey, newValue := range newDeployment.Spec.Template.ObjectMeta.Annotations {
+		oldValue, ok := existingDeployment.Spec.Template.ObjectMeta.Annotations[newKey]
+		if !ok || oldValue != newValue {
+			metadataChanged = true
+			updatedAnnotations[newKey] = newValue
+		}
+	}
+
+	// Fields intentionally ignored:
+	//    Spec.Replicas - Since a HorizontalPodAutoscaler explicitly changes this value on the deployment directly
+
+	// Verify everything in the container spec aside from some particular exceptions (see "ignoreFields" below)
+	existingContainerCount := len(existingDeployment.Spec.Template.Spec.Containers)
+	if existingContainerCount == 0 {
+		// This is unlikely but if it happens, replace the entire existing deployment with a proper one
+		logger.Error("Deployment Has No Containers")
+		return newDeployment, true
+	} else if existingContainerCount > 1 {
+		logger.Warn("Deployment Has Multiple Containers; Comparing First Only")
+	}
```

Comment:
Unless there's some K8S standard guaranteeing this we need to find the right container.  Istio Injected sidecars, etc...

Reply:
I've added a guard (findContainer) to ensure that the old container name is the same as the new once.  Istio (and, I presume, other) sidecar injectors don't modify the deployment, though; they only add containers to the pods themselves.
Other things do modify the deployment, like HorizonalPodAutoscalers (directly changes the Replicas), though, and something unknown might add more containers to it, so I think this was a good call regardless.